### PR TITLE
Origin/feature/bbb 19 reset other users password

### PR DESCRIPTION
**Pull request by pauljroberts**

### CODE REVIEW

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/client/usermanagement/UserManagementApiClient.java [hunk 1]
```diff
@@ -161,9 +159,17 @@ public void deleteUser(Integer localAuthorityId, Integer userId) {
     restTemplateFactory.getInstance().delete(uri, localAuthorityId, userId);
   }
 
-  public void resetPassword(Integer authorityId, Integer userId) {
-    // TODO mocked out API.  To be replaced.
-    logger.warn("Using mock resetPassword api.  To be implemented.");
+  public void requestResetPassword(Integer localAuthorityId, Integer id) {
```
Comment (mgilgar) [inline src/main/java/uk/gov/dft/bluebadge/webapp/la/client/usermanagement/UserManagementApiClient.java]: I would call it requestPasswordReset as per the endpoint "/authorities/{authorityId}/users/{userId}/passwordReset"
or change the endpoint

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java [hunk 2]
```diff
@@ -23,18 +23,15 @@
...
-
-  public static final String PARAM_ID = "id";
-
-  public static final String MODEL_FORM_REQUEST = "formRequest";
-  public static final String MODEL_ID = "id";
-
+  private static final String PARAM_ID = "id";
+  private static final String MODEL_FORM_REQUEST = "formRequest";
+  private static final String MODEL_ID = "id";
+  private static final String URL_REQUEST_RESET_PASSWORD =
```
Comment (mgilgar) [inline src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java]: I would call it URL_REQUEST_PASSWORD_RESET

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java [hunk 3]
```diff
@@ -111,4 +108,22 @@ private User combine(final UserDetailsFormRequest formRequest, final UserData us
...
     return user;
   }
+
+  @PostMapping(URL_REQUEST_RESET_PASSWORD)
+  public String requestPasswordReset(
+      @PathVariable(PARAM_ID) int id,
+      @ModelAttribute(MODEL_FORM_REQUEST) UserDetailsFormRequest formRequest,
+      Model model) {
+    try {
+      userService.requestResetPassword(formRequest.getLocalAuthorityId(), id);
```
Comment (mgilgar) [inline src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java]: .requestPasswordReset(

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java [hunk 4]
```diff
@@ -111,4 +108,22 @@ private User combine(final UserDetailsFormRequest formRequest, final UserData us
...
+      return REDIRECT_URL_MANAGE_USERS;
+    } catch (Exception ex) {
```
Comment (mgilgar) [inline src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java]: I think we can remove this code, if there is a 500, it should be caught soon by a controller advice. If you do not have a requirement in the story to do this, I think you can remove it.

File: src/main/resources/templates/manage-users/user-details.html [hunk 5]
```diff
@@ -23,51 +23,51 @@ <h1 class="govuk-heading-l" data-uipath="title">User details</h1>
 
                     <form th:action="@{/manage-users/user-details/{id}(id=${id})}" th:object="${formRequest}" method="post" novalidate>
 
-                <div th:replace="~{ fragments/form-elements :: govuk-input('Full name', 'name', null, 'text') }"/>
+                <div th:replace="~{ fragments/form-elements :: govuk-input(#{userdetail.field.fullname}, 'name', null, 'text') }"></div>
```
Comment (mgilgar) [inline src/main/resources/templates/manage-users/user-details.html]: Even though it is actually the full name, I would call it name to match the id of the input. That way when we see the message in message.properties, we know it should be use in "name" fields.

File: src/main/resources/templates/manage-users/user-details.html [hunk 6]
```diff
@@ -23,51 +23,51 @@ <h1 class="govuk-heading-l" data-uipath="title">User details</h1>
...
 
-                <div th:replace="~{ fragments/form-elements :: govuk-input('Email address', 'emailAddress', null, 'email') }"/>
+                <div th:replace="~{ fragments/form-elements :: govuk-input(#{userdetail.field.email}, 'emailAddress', null, 'email') }"></div>
```
Comment (mgilgar) [inline src/main/resources/templates/manage-users/user-details.html]: I would call it "emailAddress" to match the id of the input.

File: src/main/resources/templates/manage-users/user-details.html [hunk 7]
```diff
@@ -23,51 +23,51 @@ <h1 class="govuk-heading-l" data-uipath="title">User details</h1>
...
                 <summary class="govuk-details__summary">
-                    <span class="govuk-details__summary-text" data-uipath="resetPasswordSummaryText">Reset Password</span>
+                    <span class="govuk-details__summary-text" data-uipath="resetPasswordSummaryText" th:text="#{userdetail.resetPasswordSummaryText}"></span>
                 </summary>
 
                 <div class="govuk-details__text">
-                    <form th:action="@{/manage-users/user-details/{id}(id=${id})}" th:object="${formRequest}" th:method="delete">
-                        <p>Are you sure you want to reset the password for <strong data-uipath="reset.userName" th:text="*{name} + '?'" /></p>
-                        <p>We'll send them an email to reset their password.</p>
+                    <form th:action="@{/manage-users/request-reset-password/{id}(id=${id})}" th:object="${formRequest}" th:method="post">
```
Comment (mgilgar) [inline src/main/resources/templates/manage-users/user-details.html]: if you change the URL in the controller, this should be changed to /request-password-reset

File: src/main/resources/templates/manage-users/user-details.html [hunk 8]
```diff
@@ -23,51 +23,51 @@ <h1 class="govuk-heading-l" data-uipath="title">User details</h1>
...
                     <form th:action="@{/manage-users/user-details/{id}(id=${id})}" th:object="${formRequest}" method="post" novalidate>
 
-                <div th:replace="~{ fragments/form-elements :: govuk-input('Full name', 'name', null, 'text') }"/>
+                <div th:replace="~{ fragments/form-elements :: govuk-input(#{userdetail.field.fullname}, 'name', null, 'text') }"></div>
 
-                <div th:replace="~{ fragments/form-elements :: govuk-input('Email address', 'emailAddress', null, 'email') }"/>
+                <div th:replace="~{ fragments/form-elements :: govuk-input(#{userdetail.field.email}, 'emailAddress', null, 'email') }"></div>
 
-                <input value="Update" type="submit" data-uipath="updateUserButton" class="govuk-button">
+                <input th:value="#{common.update.button}" type="submit" data-uipath="updateUserButton" class="govuk-button">
```
Comment (mgilgar) [inline src/main/resources/templates/manage-users/user-details.html]: In some other messages, i.e.:
userdetail.text.confirmpassword.line2
you are using the convention of page.typeOfElement.IdOrName
But here you use common which sounds right but then update.button, which is, hard to explain, probably should be common.button.update.

Do you agree? Or Am I missing something else?

File: src/test/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsControllerTest.java [hunk 9]
```diff
@@ -201,4 +200,17 @@ public void deleteUser_shouldRedirectToManageUsers_WhenYouAreSignedInAndThereAre
         .andExpect(redirectedUrl(URL_MANAGE_USERS));
     verify(userServiceMock, times(1)).delete(LOCAL_AUTHORITY, USER_ID);
   }
+
+  @Test
+  public void
+      requestPasswordReset_shouldRedirectToManageUsers_WhenYouAreSignedInAndThereAreNoErrors()
```
Comment (mgilgar) [inline src/test/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsControllerTest.java]: I think the text segment "WhenYouAreSignedIn" is not longer relevant as we do not need to be signed in or signed out to test this in the standalone mode. It seems spring security decouples all of that for us.

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/client/usermanagement/UserManagementApiClient.java (hunk 1 above)
Comment (pauljroberts) [reply src/main/java/uk/gov/dft/bluebadge/webapp/la/client/usermanagement/UserManagementApiClient.java]: Done.

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java (hunk 2 above)
Comment (pauljroberts) [reply src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java]: Done

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java (hunk 3 above)
Comment (pauljroberts) [reply src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java]: Done

File: src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java (hunk 4 above)
Comment (pauljroberts) [reply src/main/java/uk/gov/dft/bluebadge/webapp/la/controller/UserDetailsController.java]: Done

File: src/main/resources/templates/manage-users/user-details.html (hunk 5 above)
Comment (pauljroberts) [reply src/main/resources/templates/manage-users/user-details.html]: Done.

File: src/main/resources/templates/manage-users/user-details.html (hunk 6 above)
Comment (pauljroberts) [reply src/main/resources/templates/manage-users/user-details.html]: Done

File: src/main/resources/templates/manage-users/user-details.html (hunk 7 above)
Comment (pauljroberts) [reply src/main/resources/templates/manage-users/user-details.html]: Done.

File: src/main/resources/templates/manage-users/user-details.html (hunk 8 above)
Comment (pauljroberts) [reply src/main/resources/templates/manage-users/user-details.html]: Done.  common.button.update

**Pull request closed without merging**